File: _java-archive/kafka-listener/src/main/java/dev/kritchalach/kafka/service/ContractDraftService.java
package dev.kritchalach.kafka.service;

import dev.kritchalach.kafka.domain.Contract;
import dev.kritchalach.kafka.domain.ContractParties;
import dev.kritchalach.kafka.event.BookingConfirmedEvent;
import dev.kritchalach.kafka.event.ContractDraftCreatedEvent;
import dev.kritchalach.kafka.repository.ContractRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.util.UUID;

/**
 * Application Service สำหรับ Command: CreateContractDraft
 * รับ BookingConfirmedEvent → สร้าง Contract aggregate → emit ContractDraftCreated
 */
@Service
public class ContractDraftService {

    private static final Logger log = LoggerFactory.getLogger(ContractDraftService.class);

    private final ContractRepository contractRepository;

    public ContractDraftService(ContractRepository contractRepository) {
        this.contractRepository = contractRepository;
    }

    public ContractDraftCreatedEvent createContractDraft(BookingConfirmedEvent event) {
        UUID bookingId = UUID.fromString(event.getBookingId().replaceAll("\\s+", ""));
        UUID unitId = UUID.fromString(event.getUnitId().replaceAll("\\s+", ""));
        UUID customerId = UUID.fromString(event.getCustomerId().replaceAll("\\s+", ""));

        log.info("Command: CreateContractDraft — bookingId={}, unitId={}, customerId={}",
                bookingId, unitId, customerId);

        // สร้าง Value Objects
        UUID sellerId = UUID.randomUUID();
        ContractParties parties = new ContractParties(customerId, sellerId);

        // สร้าง Contract Aggregate ผ่าน factory method (Command pattern)
        UUID templateId = UUID.randomUUID(); // simulate template selection
        String fileUrl = "https://storage.realestate.com/contracts/draft-" + UUID.randomUUID() + ".pdf";

        Contract contract = Contract.createContractDraft(
                bookingId, customerId, unitId, templateId, parties, fileUrl);
        contract = contractRepository.save(contract);

        log.info("ContractDraftCreated: contractId={}, status=DRAFT", contract.getContractId());

        // สร้าง Domain Event: ContractDraftCreated
        return new ContractDraftCreatedEvent(
                contract.getContractId(),
                contract.getBookingId(),
                contract.getUnitId(),
                contract.getCustomerId(),
                contract.getStatus().name(),
                contract.getContractDraft().getFileUrl(),
                contract.getTemplateId(),
                contract.getCreatedAt(),
                contract.getContractDraft().getDraftedAt()
        );
    }
}

File: contract-service/src/service/ContractDraftService.ts
import { v4 as uuidv4 } from 'uuid';
import { Contract } from '../domain/Contract.js';
import { ContractParties } from '../domain/ContractParties.js';
import { ContractRepository } from '../repository/ContractRepository.js';
import { coerceToUuid } from '../util/idCoerce.js';

export interface BookingConfirmedEvent {
  bookingId: string;
  unitId: string;
  customerId: string;
}

export interface ContractDraftCreatedEvent {
  contractId: string;
  bookingId: string;
  unitId: string;
  customerId: string;
  status: string;
  fileUrl: string;
  templateId: string;
  createdAt: string;
  draftedAt: string;
}

/**
 * Application Service สำหรับ Command: CreateContractDraft
 * รับ BookingConfirmedEvent → สร้าง Contract aggregate → คืน ContractDraftCreated event
 */
export const ContractDraftService = {
  async createContractDraft(event: BookingConfirmedEvent): Promise<ContractDraftCreatedEvent> {
    console.log(
      `[Command] CreateContractDraft — bookingId=${event.bookingId}, unitId=${event.unitId}, customerId=${event.customerId}`,
    );

    // Coerce string codes (e.g. "PROP-001") to deterministic UUIDs
    const bookingId = coerceToUuid(event.bookingId);
    const unitId = coerceToUuid(event.unitId);
    const customerId = coerceToUuid(event.customerId);

    // VOs
    const sellerId = uuidv4();
    const parties = new ContractParties(customerId, sellerId);

    // Factory method (Aggregate Root)
    const templateId = uuidv4();
    const fileUrl = `https://storage.realestate.com/contracts/draft-${uuidv4()}.pdf`;

    const contract = Contract.createContractDraft({
      bookingId,
      customerId,
      unitId,
      templateId,
      parties,
      fileUrl,
    });
    await ContractRepository.save(contract);

    console.log(`[Domain Event] ContractDraftCreated: contractId=${contract.contractId}, status=DRAFT`);

    return {
      contractId: contract.contractId,
      bookingId: contract.bookingId,
      unitId: contract.unitId,
      customerId: contract.customerId,
      status: contract.status,
      fileUrl: contract.contractDraft!.fileUrl,
      templateId: contract.templateId,
      createdAt: contract.createdAt,
      draftedAt: contract.contractDraft!.draftedAt,
    };
  },
};
